openquake/hazardlib/site_amplification.py

```diff
@@ -16,36 +16,121 @@
+        # the other soil typologies
+        if soil is not None:
+            df = df[df['ampcode'] == soil]
+        if 'from_mag' in df.keys():
+            self.mags = numpy.unique(df['from_mag'])
+        self.df = df
+
+    @classmethod
+    def from_dframe(cls, df, soil=None):
+        """
+        :param df:
+            A :class:`pandas.DataFrame` instance
+        :param soil:
+            A string
+        :returns:
+            A :class:`openquake.hazardlib.site_amplification.AmplFunction`
+            instance
+        """
+
+        # Get IMTs
+        imts = []
+        for key in df.keys():
+            if re.search('^SA', key) or re.search('^PGA', key):
+                imts.append(key)
+
+        # Create the temporary list of lists
+        out = []
+        for _, row in df.iterrows():
+            tmp = [row['ampcode'], row['from_mag'],
+                   row['from_rrup'], row['level']]
+            for imt in imts:
+                med = row[imt]
+                std = row['sigma_'+imt]
+                out.append([tmp[0], tmp[1], tmp[2], tmp[3],
+                            imt, med, std])
+
+        # Create the dataframe
+        dtypes = {'ampcode': ampcode_dt, 'from_mag': float,
+                  'from_rrup': float, 'level': float, 'imt': str,
+                  'median': float, 'std': float}
+        columns = [k for k in dtypes]
+        df = pd.DataFrame(out, columns=columns)
+        df = df.astype(dtypes)
+
+        return AmplFunction(df, soil)
+
+    def get_mean_std(self, site, imt, iml, mag, dst):
+        """
+        :param site:
+            A string specifying the site
+        :param imt:
+            A string specifying the intensity measure type e.g. 'PGA' or
+            'SA(1.0)'
+        :param iml:
+            A float with the shaking level on rock for which we need the
+            amplification factor
+        :param mag:
+            A float with the magnitude of the rupture
+        :param dst:
+            A float defining the rupture-site distance
+        :returns:
+            A tuple with the median amplification factor and the std of the
+            logarithm
+        """
+
+        df = copy.copy(self.df)
+        df = df[(df['ampcode'] == site) & (df['imt'] == imt)]
+
+        # Filtering magnitude
+        idx = numpy.argmin((self.mags - mag) > 0)
+        df = df[df['from_mag'] == self.mags[idx]]
+
+        # Filtering distance
+        tmp_dsts = numpy.array(sorted(df['from_rrup']))
+        idx = numpy.argmin((tmp_dsts - dst) > 0)
+        df = df[df['from_rrup'] == tmp_dsts[idx]]
+
+        # Interpolating
+        median = numpy.interp(iml, df['level'], df['median'])
+        std = numpy.interp(iml, df['level'], df['std'])
+
+        return median, std
+
+    def get_max_sigma(self):
+        """
+        :returns:
+            The maximum sigma value in the amplification function
+        """
+        # TODO - this must be improved
+        # return max(self.df['std'])
```

Comment:
Not clear why the 0.3 is hard coded here and why `max(self.df['std'])` is commented out

Reply:
Fixed